### PR TITLE
Refactor POST_friend and show mods and mod invites their permissions / permission changes.

### PR DESCRIPTION
**Pull request by 13steinj**
This PR fixes duplicate permbans/permwikibans, refactors POST_friend to be more efficient or return proper errors when needed, sends messages to those that become wikicontributors/wikibanned, sends a message to mods when they are demodded / their permissions altered, and shows moderator invites their permissions under the acceptance question so they know what they are getting into. 

Edit: if 350338c of this PR is accepted, 991abc3 of #1466 is not needed.

References:

* https://www.reddit.com/r/bugs/comments/3o9p3c/when_banning_adding_someone_to_the_wiki_as_a/?ref=search_posts

* https://www.reddit.com/r/bugs/comments/3o9oms/a_user_can_be_banned_from_a_subreddits_wiki_and/?ref=search_posts

* https://www.reddit.com/r/bugs/comments/3p2h8l/banned_twice_and_user_got_sent_two_messages/?ref=search_posts

* https://www.reddit.com/r/ModSupport/comments/3seyzg/if_a_moderators_permissions_are_changed_have_a_pm/?ref=search_posts

### CODE REVIEW

File: r2/r2/controllers/api.py [hunk 1]
```diff
@@ -885,6 +908,10 @@ def POST_unfriend(self, nuser, iuser, container, type):
         if new and type in self._sr_friend_types:
             ModAction.create(container, c.user, action, target=victim)
 
+        if was_mod:
+            send_mod_permission_message(container, c.user, victim,
+                                        was_mod_perms, demod=True)
```
Comment (Zazie-Lavender) [inline r2/r2/controllers/api.py]: I do have concerns about abuse.

Possibly with two processes spanning two accounts on a mod list.

https://www.reddit.com/r/ideasfortheadmins/comments/42h0l3/get_a_notification_when_you_are_demodded_from_a/czaege0

Unless I'm not seeing something crucial here like the API limits being a limiting factor...

File: r2/r2/lib/system_messages.py [hunk 2]
```diff
@@ -130,10 +145,70 @@ def send_ban_message(subreddit, mod, user, note=None, days=None, new=True):
...
         "[site rules](/rules) and can result in being banned from reddit "
-        "entirely.")
+        "entirely." % {"ban_type": ban_type})
 
     item, inbox_rel = Message._new(mod, user, subject, message, request.ip,
         sr=subreddit, from_sr=True)
     queries.new_message(item, inbox_rel, update_modmail=False)
+
+
+def send_mod_permission_message(subreddit, changer, mod, oldperms,
```
Comment (Zazie-Lavender) [inline r2/r2/lib/system_messages.py]: Is there perhaps a way to check if the one being demoted is banned/muted before sending a notification? (or perhaps adding a tool for an admin to silently adjust permissions, if it doesn't already exist)

File: r2/r2/controllers/api.py (hunk 1 above)
Comment (13steinj) [inline r2/r2/controllers/api.py]: I don't understand, what is there to abuse? The sending of messages? There is a limit to how often permissions can be changed as well as how many mods can be added at a time. Not sure about demodding, but it gets balanced out by the other two?

File: r2/r2/lib/system_messages.py (hunk 2 above)
Comment (13steinj) [inline r2/r2/lib/system_messages.py]: You can't ban / mute a mod; they have to be demodded first.

Admins can easily silently adjust permissions via the CLI interpreter, but they rarely get involved in subs anyway unless something bad happens, and when that does happen, they deserve a notification i'd assume.

Comment (Zazie-Lavender) [inline r2/r2/lib/system_messages.py]: Oh good! Then that resolves my worries; I suspected admins had some tools that weren't immediately apparent. :+1:  Thank you for your swift reply!

Comment (13steinj) [inline r2/r2/lib/system_messages.py]: Np. Also, I didn't see your ninja edit, so to answer to that, people that are suspended can not demod others, nor can they change permissions or add mods. A person who is suspended can have their perms changed by another mod, or be demodded, or even accept the mod invite IIRC, but that doesn't really matter.

**Comment by umbrae** [comment]
Hey @13steinj, thanks for your efforts as always. I think this is honestly too large for us to feel like we can confidently triage and review it and say with confidence these are things we want/need and are done the right way.

In the future, I think if you'd like something merged you should check with us first before starting and we can try to help, or try small patches that are easier to digest. I hope that makes sense.

**Comment by 13steinj** [comment]
@umbrae with all due respect, while I understand that reviewing something of this size can seem daunting, I've seen much larger changes, reviewed by teams far smaller that of reddit (even a 1 man job).

That said, I don't do this so they're merged. I do it because 1, I like to tinker, and 2, I want to show the people that are requesting these things on /r/ModSupport and /r/ideasfortheadmins and wherever else that it's far more highly possible to do than the majority of them think, or of they think it is very possible, they assume it would take weeks, months to do when in reality, it takes far less than that (of course, that's coming from an individual who's not jumping through hoops of UI/UX/management teams, but still).

I have no problem with you closing this. Hell, I have no problem if you close literally every PR I make, or even set up a bot to do so, or somehow ban me from making them (if that's a thing that can be done on github, I don't even know).

However, I have a very big problem with what you just said, for a number of reasons. Firstly, in a way, I used to check with you guys. Well, one of you specifically. @Deimos reached out to me a long time ago and asked me to do the same thing because he didn't want me to waste my time. And I did, ask him that is, on numerous things. And I stopped doing that for three reasons. One, I was uncomfortable doing so, I felt as if I was being a bother. Two, it kinda limited what I was doing and I was doing what I was and still am doing because I'm very, very bored on my days. Limiting what I want to do and make doesn't help alleviate that boredom. And three, even what I was doing sometimes didn't get accepted (which is fully understandable because he might have been speaking on what he, personally thought wouldn't be a waste of my time without asking all of your teams), and as I said before, acceptance wasn't really something I cared about so combined with reason two it didn't make sense to continue.

On top of that, you're saying small patches, but...well, even some of the small things I do aren't done. Hell, I'll give you the benefit of the doubt and say you somehow forgot about #1451 twice. There's another small PR that's open which was left hanging by the fact that two people the reviewer @mentioned didn't respond, and now, the original reviewer doesn't even work at reddit anymore. There's another small-medium one left hanging for reasons I don't currently know, I'm just gonna guess that Harold didn't pick up the haircut, which that said I'm also at fault since I never double haircutted and I still haven't since my virtual machine broke around a week ago and that's the one I can work from most easily. I'm going to assume you mean bugfixes alone, which is definitely understandable, but like I said that's kinda limiting me. And, I don't really want to bother and pester you guys with everything I find. 

But also, I'm highly disappointed in the level of admin response in /r/modsupport. It seems you guys don't even check the sub, since a piece of spam was removed a whole two hours after it was posted, and not because of casual checking, but because I specifically pinged redtaboo. And if you go and sort by top of all time, and skip the admin posts, you see that as time went on and people stopped noticing, the level of admin response to even relatively highly rated suggestions just...stopped. Even some of the things that I did since they were actually quick to do. The most personally painful occurrence of this was with [this ordeal](https://www.reddit.com/r/ModSupport/comments/3uwr89/i_would_like_to_see_a_worst_sort_for_comments/), because there (supposedly) *was* an admin response. However, it wasn't public. /u/Motha_Effin_Kitty_Yo who was the original requester had told me that he messaged you guys and was told there will be a response in a few days. But instead of going into the thread I linked, and responding there, to even try to calm the well of semi outrage and semi excitedness, MEKY was replied to privately and was kind enough to relay the message to me (which, side note that I'm not going into atm since I could be very, very wrong, but I didn't fully believe those reasons then, and I believed them even less later after seeing some interesting commits that caught my eye). If MEKY isn't a full blown liar, I'm just saying that I personally, among probably others, are highly disappointed.

**Pull request closed without merging**